Add UniFi Uptime sensor

timkoers opened the pull request:
Added the UniFi uptime data as a sensor. Untested.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
New sensor feature, the UniFi Uptime time.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-asc+-review%3Aapproved

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

timkoers committed:
Added UniFi Uptime sensor

Added the UniFi uptime data as a sensor. Untested.

timkoers committed:
Update sensor.py

Updated code as a result of the tests.

Kane610 commented:
You should put this in draft state until it is ready for review

MartinHjelmare commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -100,3 +101,25 @@ def state(self) -> int:
+    @property
+    def unit_of_measurement(self) -> str:
+        """Return the unit of measurement of this entity."""
+        return TIME_SECONDS
```
This should be a timestamp sensor instead without unit.
https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

MartinHjelmare commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -100,3 +101,25 @@ def state(self) -> int:
+    @property
+    def state(self) -> int:
+        """Return the uptime of the client."""
+        return self.client.uptime
```
The state should be UTC absolute isoformatted timestamp.

timkoers committed:
Changed timestamp format and device class

Converted state to iso timestamp and changed device class to DEVICE_CLASS_TIMESTAMP.

MartinHjelmare commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -100,3 +104,30 @@ def state(self) -> int:
+        """Return the name of the client."""
+        return f"{super().name} {self.TYPE.upper()}"
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Return the unit of measurement of this entity."""
+        return ""
+
```
The default unit is `None`, which is correct here.
```suggestion
```

MartinHjelmare commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -100,3 +104,30 @@ def state(self) -> int:
+    @property
+    def state(self) -> int:
+        """Return the uptime of the client."""
+        return dt_util.utc_from_timestamp(float(self.client.uptime))
```
```suggestion
        return dt_util.utc_from_timestamp(float(self.client.uptime)).isoformat()
```

timkoers committed:
Updated unit of measurement to None

MartinHjelmare commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -100,3 +104,30 @@ def state(self) -> int:
+        """Return the name of the client."""
+        return f"{super().name} {self.TYPE.upper()}"
+
+    @property
+    def unit_of_measurement(self) -> str:
+        """Return the unit of measurement of this entity."""
+        return None
+
```
```suggestion
```

timkoers committed:
Added import

timkoers committed:
Update homeassistant/components/unifi/sensor.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

Kane610 commented:
Other things expected prior to approval:
Add option in options flow to enable/disable this
Add option property to controller class with default being disabled 
Add tests

MartinHjelmare commented:
If it's a single sensor the enabling/disabling can be done via the system options already.

Kane610 commented:
It's per client so potentially a lot of entities

timkoers committed:
Removed whitespace

timkoers committed:
Added the uptime sensors option to the config flow

timkoers committed:
All the unit tests should be there now

timkoers committed:
Whoops

timkoers committed:
Fixed translation

timkoers committed:
Properly formatted the code

timkoers committed:
Flake8 really has angel eyes

timkoers commented:
> 
> 
> Other things expected prior to approval:
> Add option in options flow to enable/disable this
Done
> Add option property to controller class with default being disabled
Done
> Add tests
Done

timkoers committed:
Black should also be satisfied now

timkoers committed:
Should have satisfied all static code analysis tools

Kane610 commented:
I'll have a deep dive tonight, starting to look good!

timkoers committed:
Fixed add uptime sensor function

timkoers committed:
Fixed overintendation

timkoers commented:
Is it useful to have a `wireless_client_name_uptime `and a `wired_client_name_uptime`? Or should I just make one uptime per client?

Kane610 commented:
They usually have different Mac addresses so you can't really know, a wired device should never be seen as a wireless device and vice versa, except due to this old UniFi bug

timkoers committed:
Fixed unit tests

timkoers commented:
Oh wait, I've got it. I got a little confused by the hostnames of the 'clients' inside the unit tests

timkoers committed:
Made a spelling mistake during editing of unit tests

timkoers committed:
Test verifies if utc time is correct

timkoers committed:
Converted to iso format

timkoers committed:
Converted unit test to iso format

timkoers committed:
Unit test sensor json had the wrong uptime name

timkoers committed:
Added options_updated handler

timkoers committed:
Fixed remove sensors unit test

timkoers commented:
@MartinHjelmare I'm not sure which change is still open for request, but I've fixed every request you made. I did some of them locally (not clicking the commit request button on github)

MartinHjelmare commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -55,6 +64,22 @@ def add_entities(controller, async_add_entities, clients):
 
 
+@callback
+def add_uptime_entities(controller, async_add_entities, clients):
```
Maybe try and reuse `add_entities` callback by adding a parameter for the type of entity? We can create a dict that maps the entity type to the entity class.

Kane610 replied:
I think it is fine like this, it mirrors all other platforms in UniFi integration, makes it easier to navigate

MartinHjelmare commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -100,3 +125,31 @@ def state(self) -> int:
+    """UniFi uptime sensor base class."""
+
+    DOMAIN = DOMAIN
+    CLASS = DEVICE_CLASS_TIMESTAMP
```
Where is `CLASS` used?

timkoers replied:
The device class needed to be the timestamp class right?

MartinHjelmare replied:
We only need to set the `device_class` property correctly which is already done. We can remove this.

MartinHjelmare replied:
Please remove this.
```suggestion
```

timkoers commented on tests/components/unifi/test_sensor.py:
```diff
@@ -134,43 +159,61 @@ async def test_sensors(hass):
     wired_client_tx = hass.states.get("sensor.wired_client_name_tx")
     assert wired_client_tx is not None
 
+    wired__client_uptime = hass.states.get("sensor.wired_client_name_uptime")
```
I thought I fixed them

Kane610 replied:
Nope, still there

MartinHjelmare replied:
Bump.

timkoers committed:
Update homeassistant/components/unifi/sensor.py

Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

timkoers committed:
Update homeassistant/components/unifi/sensor.py

Co-authored-by: Robert Svensson <Kane610@users.noreply.github.com>

timkoers committed:
Update test_device_tracker.py

Removed uptime from the devices

timkoers committed:
Fixed black formatting issue

timkoers commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -100,3 +125,31 @@ def state(self) -> int:
+    """UniFi uptime sensor base class."""
+
+    DOMAIN = DOMAIN
+    CLASS = DEVICE_CLASS_TIMESTAMP
```
So both are not needed, or is only the class not needed?

Kane610 replied:
CLASS and DEVIXE_CLASS you've added the property which is the correct way of doing it

timkoers replied:
Are you sure? The unit tests now fail.

timkoers commented:
@MartinHjelmare and @Kane610 I'm not sure which changes are still open for request, but I've fixed every request you made. I did some of them locally (not clicking the commit request button on github)

Kane610 commented:
Nearly there! Smaller things left. Please also add a test to comply with the coverage comment; verify that the same client gets added twice

timkoers committed:
I think the code coverage should be good now

timkoers committed:
Trying to add the sensors again

timkoers committed:
Using signals to hopefully trigger the controller to add them again

timkoers committed:
Forgot import

timkoers committed:
Sorted components

timkoers committed:
fixed isort comments

timkoers commented:
> 
> 
> Nearly there! Smaller things left. Please also add a test to comply with the coverage comment; verify that the same client gets added twice

I've added the test that covers the uncovered line, including homeassistant/components/unifi/sensor.py#L58.

timkoers committed:
Removed CLASS and DEVICE_CLASS

MartinHjelmare commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -131,8 +131,6 @@ class UniFiUpTimeSensor(UniFiClient):
 
     DOMAIN = DOMAIN
-    CLASS = DEVICE_CLASS_TIMESTAMP
-    TYPE = UPTIME_SENSOR
```
`TYPE` needs to stay.

timkoers committed:
Added TYPE again

timkoers reviewed (commented):
👍

timkoers committed:
Removed double underscores

MartinHjelmare reviewed (approved):
Looks good!

Code owner should approve before merge.

MartinHjelmare commented:
Docs should be updated:
https://www.home-assistant.io/integrations/unifi/#sensor

timkoers reviewed (commented):
All good, all of them have been fixed

Kane610 reviewed (approved):
Awesome! Good job
